File: app/src/main/kotlin/com/openclaw/ghostcrab/ui/onboarding/OnboardingViewModel.kt
package com.openclaw.ghostcrab.ui.onboarding

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openclaw.ghostcrab.domain.model.OnboardingStep
import com.openclaw.ghostcrab.domain.repository.OnboardingRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * ViewModel for the onboarding walkthrough.
 *
 * Persists progress via [OnboardingRepository] so the user can resume after backgrounding.
 * Navigation between steps is linear; [OnboardingStep.Completed] is the terminal state.
 *
 * @param repository Stores and retrieves onboarding progress.
 */
public class OnboardingViewModel(
    private val repository: OnboardingRepository,
) : ViewModel() {

    private val _step = MutableStateFlow<OnboardingStep>(OnboardingStep.Welcome)

    /** Current onboarding step. Emits [OnboardingStep.Completed] when done. */
    public val step: StateFlow<OnboardingStep> = _step.asStateFlow()

    init {
        viewModelScope.launch {
            val saved = repository.getSavedStep()
            if (saved != OnboardingStep.Completed) _step.value = saved
        }
    }

    /** Advances to the next step and persists the change. */
    public fun next() {
        viewModelScope.launch { advance(forward = true) }
    }

    /** Returns to the previous step and persists the change. No-op at [OnboardingStep.Welcome]. */
    public fun back() {
        viewModelScope.launch { advance(forward = false) }
    }

    /** Skips directly to completion, marking onboarding done. */
    public fun skip() {
        viewModelScope.launch { complete() }
    }

    /**
     * Called when the user successfully connects to a gateway during onboarding.
     * Marks onboarding as completed.
     */
    public fun markOnboardingConnected() {
        viewModelScope.launch { complete() }
    }

    private suspend fun advance(forward: Boolean) {
        val current = _step.value
        val next = when {
            forward -> when (current) {
                OnboardingStep.Welcome -> OnboardingStep.WhatIsOpenClaw
                OnboardingStep.WhatIsOpenClaw -> OnboardingStep.InstallGateway
                OnboardingStep.InstallGateway -> OnboardingStep.StartGateway
                OnboardingStep.StartGateway -> OnboardingStep.VerifyRunning
                OnboardingStep.VerifyRunning -> OnboardingStep.FindOnNetwork
                OnboardingStep.FindOnNetwork -> OnboardingStep.Completed
                OnboardingStep.Completed -> OnboardingStep.Completed
            }
            else -> when (current) {
                OnboardingStep.Welcome -> OnboardingStep.Welcome
                OnboardingStep.WhatIsOpenClaw -> OnboardingStep.Welcome
                OnboardingStep.InstallGateway -> OnboardingStep.WhatIsOpenClaw
                OnboardingStep.StartGateway -> OnboardingStep.InstallGateway
                OnboardingStep.VerifyRunning -> OnboardingStep.StartGateway
                OnboardingStep.FindOnNetwork -> OnboardingStep.VerifyRunning
                OnboardingStep.Completed -> OnboardingStep.FindOnNetwork
            }
        }
        _step.value = next
        repository.saveStep(next)
    }

    private suspend fun complete() {
        _step.value = OnboardingStep.Completed
        repository.markCompleted()
    }

    public companion object {
        /**
         * Generates a cryptographically random 32-byte token encoded as URL-safe Base64.
         *
         * @return A 43-character base64url string (no padding) suitable for use as a bearer token.
         */
        public fun generateToken(): String {
            val bytes = ByteArray(32)
            java.security.SecureRandom().nextBytes(bytes)
            return java.util.Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)
        }
    }
}

File: ios/GhostCrab/UI/Onboarding/OnboardingViewModel.swift
import Foundation
import Observation

/// ViewModel for the multi-step onboarding walkthrough.
///
/// Mirrors `OnboardingViewModel.kt`. Persists progress through `OnboardingRepository`
/// so the user can resume after backgrounding. Navigation between steps is linear;
/// `.completed` is the terminal state that signals the host to navigate to the
/// connection picker.
@MainActor
@Observable
public final class OnboardingViewModel {

    // MARK: - Dependencies

    private let repository: any OnboardingRepository

    // MARK: - State

    /// The current visible step. Starts at `.welcome`; hydrated from repository on load.
    public private(set) var step: OnboardingStep = .welcome

    /// `true` once `markCompleted()` has finished running. The host view observes
    /// this and pushes `.connectionPicker` (matching the Android `onDone` callback).
    public private(set) var isFinished: Bool = false

    // MARK: - Init

    public init(repository: any OnboardingRepository) {
        self.repository = repository
        Task { await self.loadSavedStep() }
    }

    // MARK: - Actions

    /// Advances to the next step in the linear sequence, persisting the change.
    /// At `.findOnNetwork`, `next()` triggers completion.
    public func next() {
        Task {
            let target = forward(from: self.step)
            self.step = target
            await self.repository.saveStep(target)
            if target == .completed { await self.finish() }
        }
    }

    /// Returns to the previous step, persisting the change. No-op at `.welcome`.
    public func back() {
        Task {
            let target = backward(from: self.step)
            self.step = target
            await self.repository.saveStep(target)
        }
    }

    /// Skips directly to completion. Marks onboarding done.
    public func skip() {
        Task { await self.finish() }
    }

    /// Marks onboarding as completed in repository and flips `isFinished`.
    private func finish() async {
        self.step = .completed
        await self.repository.markCompleted()
        self.isFinished = true
    }

    // MARK: - Hydration

    private func loadSavedStep() async {
        let saved = await self.repository.getSavedStep()
        if saved != .completed { self.step = saved }
    }

    // MARK: - Pure transitions (mirror Kotlin `advance()` blocks)

    private func forward(from current: OnboardingStep) -> OnboardingStep {
        switch current {
        case .welcome:        return .whatIsOpenClaw
        case .whatIsOpenClaw: return .installGateway
        case .installGateway: return .startGateway
        case .startGateway:   return .verifyRunning
        case .verifyRunning:  return .findOnNetwork
        case .findOnNetwork:  return .completed
        case .completed:      return .completed
        }
    }

    private func backward(from current: OnboardingStep) -> OnboardingStep {
        switch current {
        case .welcome:        return .welcome
        case .whatIsOpenClaw: return .welcome
        case .installGateway: return .whatIsOpenClaw
        case .startGateway:   return .installGateway
        case .verifyRunning:  return .startGateway
        case .findOnNetwork:  return .verifyRunning
        case .completed:      return .findOnNetwork
        }
    }
}
